### PR TITLE
terror: create a new error type that has class and code for comparison.

### PR DESCRIPTION
**Pull request by coocood**
The current error check can only handle errors that has exactly the same message, when error message is not constant. For example parsing error has the message describe the context of the error, we will not be able to check it.. 

With this error type, we can includes context information in the message and still able to check the error type.

### CODE REVIEW

File: terror/terror.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+		Code:    code,
+		Message: message,
+	}
+}
+
+// Error implements error interface and adds integer Class and Code, so
+// errors with different message can be compared.
+type Error struct {
+	Class   ErrClass
+	Code    ErrCode
```
Comment (shenli) [inline terror/terror.go]: Error code is enough?

**Comment by siddontang** [comment]
now we have util/errors util/errors2 and here terror, can we unify them all?

File: terror/terror.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+// ErrClass represents a class of errors.
+type ErrClass int
+
+// ErrCode represents a specific error type in a error class.
+// Same error code can be used in different error classes.
+type ErrCode int
+
+// Error classes
+const (
+	Parser ErrClass = iota
```
Comment (siddontang) [inline terror/terror.go]: iota + 1

File: terror/terror.go (hunk 1 above)
Comment (coocood) [inline terror/terror.go]: Then all error code need to be defined in one place to avoid error code confliction, it is hard to maintain.

**Comment by coocood** [comment]
@siddontang after this has been used, `util/errors` and `util/errors2` can be abandoned.

**Comment by qiuyesuifeng** [comment]
Seems not, util/errors2 has error equal check function, it is better to unify them.

File: terror/terror.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+// EqualCode returns true if err is *Error with the same clase and code.
+func (ec ErrClass) EqualCode(err error, code ErrCode) bool {
+	e := errors.Cause(err)
+	if e == nil {
+		return false
+	}
+	if te, ok := e.(*Error); ok {
+		return te.Class == ec && te.Code == code
```
Comment (qiuyesuifeng) [inline terror/terror.go]: Any need to check `te.Class == ec`?

File: terror/terror.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+	return false
+}
+
+// NotEqualCode returns true if err is not *Error with the same class
+// and the same code.
+func (ec ErrClass) NotEqualCode(err error, code ErrCode) bool {
+	return !ec.EqualCode(err, code)
+}
+
+// EqualClass return true if err is *Error with the same class.
```
Comment (qiuyesuifeng) [inline terror/terror.go]: returns

File: terror/terror.go [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+	if e == nil {
+		return false
+	}
+	if te, ok := e.(*Error); ok {
+		return te.Class == ec
+	}
+	return false
+}
+
+// NotEqualClass return true if err is not *Error with the same class.
```
Comment (qiuyesuifeng) [inline terror/terror.go]: returns

File: terror/terror.go [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+// Error classes
+const (
+	Parser ErrClass = iota + 1
+	Optimizer
+	KV
+	Server
+	// Add more as needed.
+)
+
+// String implements Stringer interface.
```
Comment (qiuyesuifeng) [inline terror/terror.go]: fmt.Stringer

File: terror/terror.go (hunk 3 above)
Comment (coocood) [inline terror/terror.go]: Different class can have the same error code, but not the same error.

**Comment by siddontang** [comment]
no function to check two errors in different ErrClass?

**Comment by coocood** [comment]
@siddontang The use case for this package is to check if one error is a specific type, I don't see strong need to compare two errors in different ErrClass.

Or did you mean to move the function `ErrorEqual` and `ErrorNotEqual` in `util/errors2` to this pacakge? Then we can do it later after this PR has been merged.

**Comment by siddontang** [comment]
LGTM

**Pull request merged**